orchestrate-containers/collab.md

```diff
@@ -141,6 +141,201 @@ Lastly, to finish up the configuration of the master node, we need to add the ne
 kubectl apply -f https://raw.githubusercontent.com/coreos/flannel/bc79dd1505b0c8681ece4de4c0d86c5cd2643275/Documentation/kube-flannel.yml
 ```
 
+---
+## Centralised Logging
+
+To monitor individual nodes can be a hectic job. With a centralised logger, it easy for administrator to monitor all the nodes. One of the common logging stack in kubernetes is Elasticsearch, Fluentd and Kibana aka EFK.
+
+EFK is comprised of:
+
+* **Elasticsearch** is a distributed search and analytics engine, an object store where all logs are stored.
+* **Fluentd** gathers logs from each nodes and forward it to Elasticsearch
+* **Kibana** is a web UI to visualise the Elasticsearch data
+
+Fluentd will be deployed on all the nodes in the cluster and aggregate the logs. Application in docker container writes stdout/stderr logs and stores in `/var/log/containers` directory.    
+​Thereafter, Fluentd will forward the log to the node where Elaasticsearch is hosted. Kibana will access Elasticsearch data and provide a visualisation of the logs collected.
+
+There are YAML files to configure the EFK stack in the Kubernetes GitHub [repository](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/fluentd-elasticsearch).
+
+You can either clone the whole repository down and run with `kubectl apply -f` or use `kubectl apply -f https://urloftheyaml` to run the YAML file for the EFK stack. In this case, the repository is clone to the master node.
+
+Firstly, the first YAML to run is the `es-statfulset.yaml` file. Do take note that this script does not configure the logs to be stored in a persistent volume. Thus, the logs will be erased when the pods terminates.
+
+To configure a **local** persistent volume, add the following script to the `es-statfulset.yaml` or a separate YAML file to create persistent volume. In this case, the script is added to the existing `es-statfulset.yaml` file. Add the script before the deployment of Elasticsearch.
+Change the size of the storage to your own preference. As the `es-statfulset.yaml` application is created in the namespace `kube-system`, same namespace has to be used for the PersistentVolume and PersistentVolumeClaim. The hostPath define the directory on where the logs should be stored in the node.
+```bash
+---
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  namespace: kube-system
+  name: elasticsearch-volume
+  labels:
+    type: local
```

Comment:
Did you guys try Ceph?

Reply:
Yes, we did try with Ceph. Below is the yaml file for es-statefulset.yaml.
```yaml
# RBAC authn and authz
apiVersion: v1
kind: ServiceAccount
metadata:
  name: elasticsearch-logging
  namespace: kube-system
  labels:
    k8s-app: elasticsearch-logging
    kubernetes.io/cluster-service: "true"
    addonmanager.kubernetes.io/mode: Reconcile
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: elasticsearch-logging
  labels:
    k8s-app: elasticsearch-logging
    kubernetes.io/cluster-service: "true"
    addonmanager.kubernetes.io/mode: Reconcile
rules:
- apiGroups:
  - ""
  resources:
  - "services"
  - "namespaces"
  - "endpoints"
  verbs:
  - "get"
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  namespace: kube-system
  name: elasticsearch-logging
  labels:
    k8s-app: elasticsearch-logging
    kubernetes.io/cluster-service: "true"
    addonmanager.kubernetes.io/mode: Reconcile
subjects:
- kind: ServiceAccount
  name: elasticsearch-logging
  namespace: kube-system
  apiGroup: ""
roleRef:
  kind: ClusterRole
  name: elasticsearch-logging
  apiGroup: ""
---
# Elasticsearch deployment itself
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: elasticsearch-logging
  namespace: kube-system
  labels:
    k8s-app: elasticsearch-logging
    version: v6.3.0
    kubernetes.io/cluster-service: "true"
    addonmanager.kubernetes.io/mode: Reconcile
spec:
  serviceName: elasticsearch-logging
  replicas: 1
  selector:
    matchLabels:
      k8s-app: elasticsearch-logging
      version: v6.3.0
  template:
    metadata:
      labels:
        k8s-app: elasticsearch-logging
        version: v6.3.0
        kubernetes.io/cluster-service: "true"
    spec:
      serviceAccountName: elasticsearch-logging
      containers:
      - image: k8s.gcr.io/elasticsearch:v6.3.0
        name: elasticsearch-logging
        resources:
          # need more cpu upon initialization, therefore burstable class
          limits:
            cpu: 1000m
          requests:
            cpu: 100m
        ports:
        - containerPort: 9200
          name: db
          protocol: TCP
        - containerPort: 9300
          name: transport
          protocol: TCP
        volumeMounts:
        - name: elasticsearch-logging
          mountPath: /data
        env:
        - name: "NAMESPACE"
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: discovery.zen.minimum_master_nodes
          value: "1" 
      volumes:
      - name: elasticsearch-logging
        cephfs:
          monitors:
            - 192.168.2.3:6789
            - 192.168.2.4:6789
            - 192.168.2.5:6789
          user: admin
          secretRef:
            name: cephfs-pass
          readOnly: false
          path: "/"
      # Elasticsearch requires vm.max_map_count to be at least 262144.
      # If your OS already sets up this number to a higher value, feel free
      # to remove this init container.
      initContainers:
      - image: alpine:3.6
        command: ["/sbin/sysctl", "-w", "vm.max_map_count=262144"]
        name: elasticsearch-logging-init
        securityContext:
          privileged: true
```

Reply:
Good. Please remember to PR the these files when it's complete.